File: Python/Exercicios_diarios/Revisão_intermediaria/Mesclar duas listas ordenadas.cpp
#include <iostream>
using namespace std;

struct ListNode
{
    int val;
    struct ListNode *next;
};

class Solution
{
public:
    ListNode *mergeTwoLists(ListNode *list1, ListNode *list2)
    {
        if (list1 == NULL)
            return list2;
        else if (list2 == NULL)
            return list1;

        if (list1->val <= list2->val)
        {
            list1->next = mergeTwoLists(list1->next, list2);
            return list1;
        }

        else
        {
            list2->next = mergeTwoLists(list1, list2->next);
            return list2;
        }
    }
};

class Solution
{
public:
    ListNode *mergeTwoLists(ListNode *list1, ListNode *list2)
    {
        if (list1 == NULL)
            return list2;
        if (list2 == NULL)
            return list1;

        ListNode *list3;
        if (list1->val <= list2->val)
        {
            list3 = list1;
            list1 = list1->next;
        }
        else
        {
            list3 = list2;
            list2 = list2->next;
        }

        ListNode *aux = list3;
        while (list1 && list2)
        {
            if (list1->val < list2->val)
            {
                aux->next = list1;
                list1 = list1->next;
            }
            else
            {
                aux->next = list2;
                list2 = list2->next;
            }
            aux = aux->next;
        }

        if (!list1) 
            aux->next = list2;
        else
            aux->next = list1;

        return list3;
    }
};

File: Python/Exercicios_diarios/Revisão_intermediaria/Mesclar duas listas ordenadas.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(self, list1, list2):
        if not list1 or not list2:
            return list1 or list2

        if list1.val < list2.val:
            list1.next = self.mergeTwoLists(list1.next, list2)
            return list1
        else:
            list2.next = self.mergeTwoLists(list1, list2.next)
            return list2


class Solution:
    def mergeTwoLists(self, list1, list2):
        list3 = aux = ListNode(0)

        while list1 and list2:
            if list1.val < list2.val:
                aux.next = list1
                list1 = list1.next
            else:
                aux.next = list2
                list2 = list2.next
                
            aux = aux.next

        aux.next = list1 or list2

        return list3.next
